crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,324 @@
+/// However, it will likely be easier to use the corresponding methods on [`World`],
+/// to avoid having to worry about split mutable borrows yourself.
+///
+/// # Limitations
+///
+///  - stored systems cannot be chained: they can neither have an [`In`](crate::system::In) nor return any values
+///  - stored systems cannot recurse: they cannot run other systems via the [`SystemRegistry`] methods on `World` or `Commands`
+///  - exclusive systems cannot be used
+///
+/// # Examples
+///
+/// You can run a single system directly on the World,
+/// applying its effect and caching its state for the next time
+/// you call this method.
+///
+/// ```rust
+/// use bevy_ecs::prelude::*;
+///
+/// let mut world = World::new();  
+///
+/// #[derive(Default, PartialEq, Debug)]
+/// struct Counter(u8);
+///
+/// fn count_up(mut counter: ResMut<Counter>){
+///     counter.0 += 1;
+/// }
+///
+/// world.init_resource::<Counter>();
+/// world.run_system(count_up);
+///
+/// assert_eq!(Counter(1), *world.resource());
+/// ```
+///
+/// These systems immediately apply commands and cache state,
+/// ensuring that change detection and [`Local`](crate::system::Local) variables work correctly.
+///
+/// ```rust
+/// use bevy_ecs::prelude::*;
+///
+/// let mut world = World::new();
+///
+/// #[derive(Component)]
+/// struct Marker;
+///
+/// fn spawn_7_entities(mut commands: Commands) {
+///     for _ in 0..7 {
+///         commands.spawn(Marker);
+///     }
+/// }
+///
+/// fn assert_7_spawned(query: Query<(), Added<Marker>>){
+///     let n_spawned = query.iter().count();
+///     assert_eq!(n_spawned, 7);
+/// }
+///
+/// world.run_system(spawn_7_entities);
+/// world.run_system(assert_7_spawned);
+/// ```
+#[derive(Resource, Default)]
+pub struct SystemRegistry {
+    systems: TypeIdMap<(bool, BoxedSystem)>,
+}
+
+impl SystemRegistry {
+    /// Registers a system in the [`SystemRegistry`], so it can be run later.
+    ///
+    /// Repeatedly registering a system will have no effect.
+    #[inline]
+    fn register<M, S: IntoSystem<(), (), M> + 'static>(&mut self, system: S) -> TypeId {
+        let type_id = TypeId::of::<S>();
+        self.systems
+            .entry(type_id)
+            .or_insert_with(|| (false, Box::new(IntoSystem::into_system(system))));
+        type_id
+    }
+
+    /// Runs the supplied system on the [`World`] a single time.
+    ///
+    /// You do not need to register systems before they are run in this way.
+    /// Instead, systems will be automatically registered according to their [`TypeId`] the first time this method is called on them.
+    ///
+    /// System state will be reused between runs, ensuring that [`Local`](crate::system::Local) variables and change detection works correctly.
+    pub fn run<M, S: IntoSystem<(), (), M> + 'static>(&mut self, world: &mut World, system: S) {
+        let id = self.register(system);
+        self.run_by_id(world, id)
+            .expect("System was registered before running");
```

Comment:
I know the rust style guide says to state expectations when using `expect` but it often leads to very unhelpful errors.
```rust
thread 'main' panicked at System was registered before running
```
Which says the exact opposite of what went wrong. Simply providing a `Debug` impl for the error type and unwrapping or just panicing with a custom message is more helpful.